### PR TITLE
Refactoring of type inference and usage

### PR DESCRIPTION
**Pull request by pluralia**
* Closes
  - #447
  - #750
  - #744
  - #583
* Improves collecting type resources for validation:
  - `LangiumGrammarDocument` extends `LangiumDocument` by adding the `validationResources` property 
  - `LangiumGrammarValidationResourcesCollector` service collects the validation resources for each document during document building
* Extracts type validations in a separate `LangiumGrammarTypesValidator`
* Adds `$type` property (discussed in #769)
* Fixes `$container` types calculation for inferred types
* Makes `types-generator.ts` (#754) using the same methods of converting to string as `ast-generator.ts`
* Reorganizes the `type-system` folder

### CODE REVIEW

File: examples/arithmetics/src/language-server/generated/ast.ts [hunk 1]
```diff
@@ -44,7 +44,7 @@ export function isBinaryExpression(item: unknown): item is BinaryExpression {
 }
 
 export interface DeclaredParameter extends AstNode {
-    readonly $container: Definition;
+    readonly $container: Definition | Module;
```
Comment (spoenemann) [inline examples/arithmetics/src/language-server/generated/ast.ts:47]: This seems wrong: a DeclaredParameter can only be contained in a Definition.

File: examples/arithmetics/src/language-server/generated/ast.ts [hunk 2]
```diff
@@ -55,7 +55,7 @@ export function isDeclaredParameter(item: unknown): item is DeclaredParameter {
 }
 
 export interface Definition extends AstNode {
-    readonly $container: Module;
+    readonly $container: Definition | Module;
```
Comment (spoenemann) [inline examples/arithmetics/src/language-server/generated/ast.ts:58]: This seems wrong: a Definition can only be contained in a Module.

File: packages/langium/src/grammar/workspace/document-builder.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import { LangiumGrammarServices } from '../langium-grammar-module';
+import { LangiumGrammarDocument } from './documents';
+
+export class LangiumGrammarDocumentBuilder extends DefaultDocumentBuilder {
+    constructor(services: LangiumSharedServices) {
+        super(services);
+        this.addTypeCollectionPhase();
+    }
+
+    private addTypeCollectionPhase() {
```
Comment (spoenemann) [inline packages/langium/src/grammar/workspace/document-builder.ts:21]: I would not do this in a subclass because there's nothing to override in this case. Proposal:
 - Move this code to a function in `langium-grammar-module.ts`
 - Call the function in `createLangiumGrammarServices`

File: packages/langium/src/test/langium-test.ts [hunk 4]
```diff
@@ -16,8 +16,9 @@ import { LangiumDocument } from '../workspace/documents';
 import { findNodeForProperty } from '../utils/grammar-util';
 import { SemanticTokensDecoder } from '../lsp/semantic-token-provider';
 import { TextDocument } from 'vscode-languageserver-textdocument';
+import { LangiumGrammarServices } from '../grammar/langium-grammar-module';
 
-export function parseHelper<T extends AstNode = AstNode>(services: LangiumServices): (input: string) => Promise<LangiumDocument<T>> {
+export function parseHelper<T extends AstNode = AstNode>(services: LangiumGrammarServices): (input: string) => Promise<LangiumDocument<T>> {
```
Comment (spoenemann) [inline packages/langium/src/test/langium-test.ts:21]: These testing utilities are not only meant for our own grammar language, but are published to by used for any Langium-based language. Please revert the changes in this file, even if it means we have to type-cast in our grammar language tests.

File: packages/langium/test/grammar/type-system/type-validator.test.ts [hunk 5]
```diff
@@ -161,3 +161,114 @@ describe('validate declared types', () => {
 
     });
 });
+
+describe('NEW TESTS', () => {
```
Comment (spoenemann) [inline packages/langium/test/grammar/type-system/type-validator.test.ts:165]: Please find a better description for this test suite.

File: packages/langium/src/grammar/generated/ast.ts [hunk 6]
```diff
@@ -200,7 +201,7 @@ export function isInferredType(item: unknown): item is InferredType {
 }
 
 export interface Interface extends AstNode {
-    readonly $container: Grammar;
+    readonly $container: Alternatives | Assignment | AtomType | CharacterRange | CrossReference | Grammar | Group | NegatedToken | ParserRule | TerminalAlternatives | TerminalGroup | TerminalRule | UnorderedGroup | UntilToken;
```
Comment (dhuebner) [inline packages/langium/src/grammar/generated/ast.ts:204]: This doesn't seem to be right. 
Can e.g. an assignment be a container of an Interface?  Other `$container` types also doesn't seem to be right to me.

**Review by spoenemann — COMMENTED** [review]
Thank you @pluralia, great step forward!

**Comment by sailingKieler** [comment]
Good job @pluralia 

I tested it with a real world grammar and ... there's an issue. Extracted it into a test case grammar:
```
grammar DeclaredReferenceTypes

entry Named: '{'
    name = ID
'}';

Expression:
    NamedRef | Literal;

Literal:
    lit = STRING
;

NamedRef:
    ref = [Named]
;

interface QualifiedRef extends NamedRef {
    qualifier: Expression
}

QualifiedRef1 infers Expression:
    NamedRef ({QualifiedRef.qualifier=current} '.' ref=[Named])+    // <-- fails with "The assigned type 'NamedRef' is not compatible with the declared property 'qualifier' of type 'Expression': , 'NamedRef' is not expected."
;

QualifiedRef2 infers Expression:
    {infer NamedRef} (ref=[Named]) ({QualifiedRef.qualifier=current} '.' ref=[Named])+    // <-- same here: "The assigned type 'NamedRef' is not compatible with the declared property 'qualifier' of type 'Expression': , 'NamedRef' is not expected."
;

// the original failing case from the real world project:
interface SpecialNamedRef extends NamedRef {}
interface SpecialQualifiedRef extends QualifiedRef {}

SpecialQualifiedRef infers Expression:
    {SpecialNamedRef} (ref=[Named]) ({SpecialQualifiedRef.qualifier=current} '.' ref=[Named])+.   // <-- fails with: "The assigned type 'SpecialNamedRef' is not compatible with the declared property 'qualifier' of type 'NamedRef': , 'SpecialNamedRef' is not expected."
;
// in the latter case, the validation message suggests that there's something wrong internally, as the declared property 'qualifier' is supposed to be of type 'Expression', not 'NamedRef'

terminal ID: /[_a-zA-Z][\w_]*/;
terminal STRING: /"[^"]*"|'[^']*'/;
``` 

You can make it even worse with an additional rule like
```
QualifiedRef3 infers Expression:
    Expression ({QualifiedRef.qualifier=current} '.' ref=[Named])+
;
```

Please play with this example, you can derive more test cases from that.

**Review by msujew — COMMENTED** [review]
This seems to reintroduce https://github.com/langium/langium/issues/559:

image

The error message also contains to many `,` characters.

There's also another issue which I've noticed about the `$type` property placement. For the following grammar (taken from the statemachine example), the following will be generated:

```
Event:
    name=ID;

interface StateRef {
    state: @State
}

interface Transition extends StateRef {
    event: @Event
}

Transition returns Transition:
    event=[Event] '=>' state=[State];
```

_generates_:

```ts
export interface Event extends AstNode {
    readonly $container: Statemachine;
    name: string
}

export interface StateRef extends AstNode {
    readonly $container: State;
    readonly $type: 'Transition';
    state: Reference<State>
}

export const StateRef = 'StateRef';

export function isStateRef(item: unknown): item is StateRef {
    return reflection.isInstance(item, StateRef);
}

export interface Transition extends StateRef {
    readonly $container: State;
    event: Reference<Event>
}
```

Note that `StateRef` correctly gets the `$type` property, while `Event` doesn't. Event should also receive the `$type: 'Event'` property.

File: packages/langium/src/grammar/generated/ast.ts (hunk 6 above)
Comment (pluralia) [reply packages/langium/src/grammar/generated/ast.ts:204]: True, it's not right by the fact: not all listed types are real `$container`s of `Interface`. However, we have them here since we collect `$container` types per connected component and not per type. We have to do it because one type can inherit some types and having `$container` types per type brings TS errors. Try an example:
```
interface A { $container: D; strA: string }
interface B { $container: E; strB: string }
interface C extends A, B { $container: D | E; strC: string }
interface D { a: A }
interface E { b: B }
```
As the result, our solution is having `$container: D | E` everywhere: for `A` and `B` to. It's not a perfect solution, of course, so, we have an issue about improving `$container` types calculation (#150). Current PR doesn't solve the problem, but adds `$container` types for declared types (#750).

Comment (msujew) [reply packages/langium/src/grammar/generated/ast.ts:204]: But `Interface` is completely isolated in it's inheritance structure. Why did it change anyway?

Comment (dhuebner) [reply packages/langium/src/grammar/generated/ast.ts:204]: Well okay, when the issue is already addressed

File: examples/arithmetics/src/language-server/generated/ast.ts (hunk 2 above)
Comment (pluralia) [reply examples/arithmetics/src/language-server/generated/ast.ts:58]: This is because `Definition` is a container type of `DeclaredParameter` and `Definition`and `DeclaredParameter` are in the same connected component.

File: packages/langium/src/grammar/generated/ast.ts (hunk 6 above)
Comment (pluralia) [reply packages/langium/src/grammar/generated/ast.ts:204]: @msujew , `Interface` is not isolated: we have `type AbstractType = Interface | Type | Action | ParserRule;` that connects it with `ParserRule` that is connected with `AbstractElement` -- what gives so many container types for each participant of this hierarchy.

**Comment by pluralia** [comment]
@sailingKieler, thank you for the wonderful example :)
The validation error that you found has already existed for parser rules before this PR (I opened an issue based on your example: #823). The reason why you didn't see this error for actions before this PR is that we had no validation for actions. This PR adds it and since it reuses code for parser rules validation, you see this error for actions, too.

File: packages/langium/test/grammar/type-system/inferred-types.test.ts [hunk 7]
```diff
@@ -889,10 +892,12 @@ describe('types of `$container` property are correct', () => {
         expect(interfacesString).toBe(s`
             export interface A extends AstNode {
                 readonly $container: D | E;
+                readonly $type: 'C';
```
Comment (msujew) [inline packages/langium/test/grammar/type-system/inferred-types.test.ts:895]: This doesn't seem to be correct. When parsing rule `A` in this grammar, we would expect an element with a `$type` with value `A` not `C`. The `$type` therefore needs to be `'A' | 'C'`.

Also, interface `D` should have `$type: 'D'` and `E` should have `$type: 'E'`.

**Comment by pluralia** [comment]
@spoenemann, this PR doesn't change container types calculation, but only adds it for declared types. So, if we don't add these changes:
1. We just get erasing `$container` property for all types that are generated from declared ones
2. We get exactly the same `$container` types calculation, like now, if we replace declared types with inferred ones. For example, if you replace a declared type `AbstractDefinition` with a parser rule, you get exactly the same container types.

Overall, upcoming changes for container types calculation are useful and helpful for the current way of calculation since it makes it more correct. I'd propose accept them or remove container types calculation at all, but not leave the current state for the release.

> because changing this back to a more constrained set of container types would be a breaking change

I don't think merging the current container types will bring us breaking changes in the future. The current way is really more generic and new way (that will be closer to real container types) is more constrained. However, extra container types that we add now, will just never be met in real life: we know that they are not real container types. So, the new way just cut what was never used.

**Comment by spoenemann** [comment]
Ok if we see the new (too large) set of container types as an issue that needs to be resolved in a 1.x version, we could merge this and regard the fix to more constrained container types as a "mildly" breaking change, meaning that it causes compile errors only in some specific situations.

File: packages/langium/src/grammar/validation/validator.ts [hunk 8]
```diff
@@ -5,31 +5,30 @@
...
     constructor(services: LangiumGrammarServices) {
         super(services);
         const validator = services.validation.LangiumGrammarValidator;
+        const typesValidator = services.validation.LangiumGrammarTypesValidator;
         const checks: ValidationChecks<ast.LangiumGrammarAstType> = {
             Action: [
-                validator.checkActionTypeUnions,
-                validator.checkAssignmentReservedName
+                validator.checkAssignmentReservedName,
+                typesValidator.checkActionIsNotUnionType,
```
Comment (spoenemann) [inline packages/langium/src/grammar/validation/validator.ts:31]: It is incorrect to mix checks from different classes in one call to `register` because it accepts only a single argument to be used as `this`. As consequence, if any check from `LangiumGrammarTypesValidator` would try to access a class property, it would fail because it would use `LangiumGrammarValidator` as `this` context. Solution: create a separate `ValidationChecks` object and call `register` a second time.

**Review by spoenemann — APPROVED** [review]
Let's ship this! From my perspective you can merge this after resolving today's comments.

File: packages/langium/src/grammar/validation/validation-resources-collector.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import { AbstractElement, Action, Grammar, Interface, isAction, isAlternatives, isGroup, isUnorderedGroup, ParserRule, Type } from '../generated/ast';
+import { getActionType, getRuleType } from '../internal-grammar-util';
+import { AstResources, collectTypeResources, TypeResources } from '../type-system/type-collector/all-types';
+import { mergeInterfaces, mergeTypesAndInterfaces } from '../type-system/types-util';
+import { Property } from '../type-system/type-collector/types';
+import { TypeToValidationInfo, ValidationResources } from '../workspace/documents';
+
+export class LangiumGrammarValidationResourcesCollector {
+
+    collectValidationResources(grammar: Grammar, documents: LangiumDocuments): ValidationResources {
```
Comment (spoenemann) [inline packages/langium/src/grammar/validation/validation-resources-collector.ts:19]: The `LangiumDocuments` could be injected in the constructor instead of passing them as parameter.
